### ISSUE TITLE
directpv v2.0.4 DaemonSet Pods are in CrashLoopBackOff

### ISSUE THREAD

**Issue by obacak** [minio/directpv#539]
```
$ kubectl-directpv install
I0304 15:51:05.774789  420871 install.go:95] Enable dynamic drive change management using --enable-dynamic-discovery flag
W0304 15:51:05.774850  420871 install.go:97] dynamic-discovery feature is strictly experimental and NOT production ready yet
I0304 15:51:05.793395  420871 ns.go:55] 'direct.csi.min.io' namespace created
I0304 15:51:05.827220  420871 rbac.go:352] 'direct.csi.min.io' rbac created
I0304 15:51:06.688334  420871 psp.go:129] 'direct.csi.min.io' podsecuritypolicy created
I0304 15:51:11.750730  420871 conversion_secret.go:170] 'direct.csi.min.io' conversion webhook secrets created
I0304 15:51:11.876521  420871 crd.go:135] crds successfully registered
I0304 15:51:11.949959  420871 csidriver.go:150] 'direct.csi.min.io' csidriver created
I0304 15:51:12.949209  420871 storageclass.go:50] 'direct.csi.min.io' storageclass created
I0304 15:51:12.968398  420871 service.go:39] 'direct.csi.min.io' service created
I0304 15:51:12.998065  420871 daemonset.go:41] 'direct.csi.min.io' daemonset created
I0304 15:51:13.036384  420871 deployment.go:295] 'direct.csi.min.io' deployment created

$ k get pod
NAME                                 READY   STATUS             RESTARTS   AGE
direct-csi-min-io-7477778cf7-2xg7v   2/2     Running            0          6s
direct-csi-min-io-7477778cf7-glfv4   1/2     Running            0          6s
direct-csi-min-io-7477778cf7-gw7bs   2/2     Running            0          6s
direct-csi-min-io-9xbwn              2/3     CrashLoopBackOff   1          6s
direct-csi-min-io-r8dq9              2/3     Error              0          6s
direct-csi-min-io-tddjm              2/3     Error              0          6s
direct-csi-min-io-vv52r              2/3     Error              1          6s
direct-csi-min-io-xhlnz              2/3     Running            0          6s

$ stern direct-csi-min-io-r8dq9
+ direct-csi-min-io-r8dq9 › liveness-probe
+ direct-csi-min-io-r8dq9 › node-driver-registrar
direct-csi-min-io-r8dq9 liveness-probe I0304 14:51:14.588279  113155 connection.go:153] Connecting to unix:///csi/csi.sock
direct-csi-min-io-r8dq9 liveness-probe W0304 14:51:24.588900  113155 connection.go:172] Still connecting to unix:///csi/csi.sock
direct-csi-min-io-r8dq9 node-driver-registrar I0304 14:51:14.286855  112927 main.go:113] Version: v2.2.0-go1.17
direct-csi-min-io-r8dq9 node-driver-registrar I0304 14:51:14.288047  112927 main.go:137] Attempting to open a gRPC connection with: "unix:///csi/csi.sock"
direct-csi-min-io-r8dq9 node-driver-registrar W0304 14:51:24.288933  112927 connection.go:172] Still connecting to unix:///csi/csi.sock
direct-csi-min-io-r8dq9 node-driver-registrar W0304 14:51:34.288299  112927 connection.go:172] Still connecting to unix:///csi/csi.sock
direct-csi-min-io-r8dq9 liveness-probe W0304 14:51:34.588706  113155 connection.go:172] Still connecting to unix:///csi/csi.sock
+ direct-csi-min-io-r8dq9 › direct-csi
direct-csi-min-io-r8dq9 direct-csi I0304 14:51:36.439757  120913 init.go:108] obtained client config successfully
direct-csi-min-io-r8dq9 direct-csi I0304 14:51:36.488759  120913 webhook.go:72] Starting conversion webhook server in port: 30443
direct-csi-min-io-r8dq9 direct-csi I0304 14:51:36.501020  120913 run.go:162] The conversion webhook is live!
direct-csi-min-io-r8dq9 direct-csi I0304 14:51:36.501068  120913 run.go:168] identity server started
direct-csi-min-io-r8dq9 direct-csi I0304 14:51:36.636192  120913 run.go:178] Enable dynamic drive change management using '--dynamic-drive-discovery' flag
direct-csi-min-io-r8dq9 direct-csi I0304 14:51:36.636206  120913 run.go:179] This flag will be made default in the next major release version
direct-csi-min-io-r8dq9 direct-csi Error: error while initializing drive discovery: read /dev/sdy: input/output error
direct-csi-min-io-r8dq9 node-driver-registrar W0304 14:51:44.288829  112927 connection.go:172] Still connecting to unix:///csi/csi.sock
direct-csi-min-io-r8dq9 liveness-probe W0304 14:51:44.588563  113155 connection.go:172] Still connecting to unix:///csi/csi.sock
```

The main problem seems to be the registration of the node-driver.

```
$ k version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.9", GitCommit:"b631974d68ac5045e076c86a5c66fba6f128dc72", GitTreeState:"clean", BuildDate:"2022-01-19T17:51:12Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.14", GitCommit:"57a3aa3f13699cf3db9c52d228c18db94fa81876", GitTreeState:"clean", BuildDate:"2021-12-15T14:47:10Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}

$ kubectl-krew version
OPTION            VALUE
GitTag            v0.4.3
GitCommit         dbfefa5
IndexURI          https://github.com/kubernetes-sigs/krew-index.git
BasePath          /home/ob52647/.krew
IndexPath         /home/ob52647/.krew/index/default
InstallPath       /home/ob52647/.krew/store
BinPath           /home/ob52647/.krew/bin
DetectedPlatform  linux/amd64

$ kubectl-directpv --version
directpv version v2.0.4
```

```
$ k get nodes -owide
NAME        STATUS   ROLES                  AGE   VERSION    INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                          KERNEL-VERSION            CONTAINER-RUNTIME
k8slabm01   Ready    control-plane,master   21d   v1.20.14   10.147.52.54    <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabm02   Ready    control-plane,master   21d   v1.20.14   10.147.52.55    <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabm03   Ready    control-plane,master   21d   v1.20.14   10.147.52.56    <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabw01   Ready    <none>                 21d   v1.20.14   10.147.52.129   <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabw02   Ready    <none>                 21d   v1.20.14   10.147.52.130   <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabw03   Ready    <none>                 21d   v1.20.14   10.147.52.131   <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabw04   Ready    <none>                 21d   v1.20.14   10.147.52.181   <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
k8slabw05   Ready    <none>                 21d   v1.20.14   10.147.52.182   <none>        Fedora CoreOS 33.20210527.epo.0   5.10.19-200.fc33.x86_64   docker://19.3.13
```

Worker node OS:
```
[root@k8slabw02 ~]# cat /etc/os-release
NAME=Fedora
VERSION="33.20210527.epo.0 (CoreOS)"
ID=fedora
VERSION_ID=33
VERSION_CODENAME=""
PLATFORM_ID="platform:f33"
PRETTY_NAME="Fedora CoreOS 33.20210527.epo.0"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:33"
HOME_URL="https://getfedora.org/coreos/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora-coreos/"
SUPPORT_URL="https://github.com/coreos/fedora-coreos-tracker/"
BUG_REPORT_URL="https://github.com/coreos/fedora-coreos-tracker/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=33
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=33
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="CoreOS"
VARIANT_ID=coreos
OSTREE_VERSION='33.20210527.epo.0'
```

I can see on a worker node that `csi.sock` for the controller was created but not for the daemonset which explains the issue but I do not know why...

```
[root@k8slabw02 ~]# ll /var/lib/kubelet/
total 44
-rw-r--r--  1 root root  1198 Feb 10 16:21 config.yaml
-rw-------  1 root root    62 Feb 10 16:21 cpu_manager_state
drwxr-xr-x  2 root root  4096 Mar  4 14:10 device-plugins
-rw-r--r--  1 root root   161 Feb 17 14:02 kubeadm-flags.env
drwxr-xr-x  2 root root  4096 Feb 10 16:22 pki
drwxr-x---  7 root root  4096 Mar  1 14:24 plugins
drwxr-x---  2 root root  4096 Mar  4 14:11 plugins_registry
drwxr-x---  2 root root  4096 Mar  4 14:10 pod-resources
drwxr-x--- 61 root root 12288 Mar  4 16:00 pods
[root@k8slabw02 ~]# ll /var/lib/kubelet/plugins
total 20
drwxr-xr-x 2 root root 4096 Mar  4 14:11 csi.hpe.com
drwxr-xr-x 2 root root 4096 Mar  1 14:24 direct-csi-min-io
drwxr-xr-x 2 root root 4096 Mar  4 15:51 direct-csi-min-io-controller
d-w-r-xr-- 3 root root 4096 Feb 10 16:28 hpe.com
drwxr-x--- 3 root root 4096 Feb 10 16:28 kubernetes.io
[root@k8slabw02 ~]# ll /var/lib/kubelet/plugins/direct-csi-min-io
total 0
[root@k8slabw02 ~]# ll /var/lib/kubelet/plugins/direct-csi-min-io-controller/
total 0
srwxr-xr-x 1 root root 0 Mar  4 15:51 csi.sock
```

For info, there is another CSI driver running in the cluster: hpe-csi v2.1.0:
```
[root@k8slabw02 ~]# ll /var/lib/kubelet/plugins_registry/
total 0
srwx------ 1 root root 0 Mar  4 14:11 csi.hpe.com-reg.sock
```
Could you please advise?

**Comment by Praveenrajmani** [comment]
cna you describe the daemonset pods alone and share the error logs here?

**Comment by Praveenrajmani** [comment]
looks like the drive `/dev/sdy` seems to be unreadable, directpv couldn't open the device for reading information. please fix the drive or remove.

please note that the next release will handle such drive failures and continue discovering other devices even if one of the drive is failing.

**Comment by obacak** [comment]
@Praveenrajmani , thank you very much !!!

**Comment by obacak** [comment]
@Praveenrajmani , any indication when this feature will be implemented? I see you have released v2.0.6 and drive failures still causes the driver failing.

Another question is; is dynamic-discovery enabled by default? We would like to have the option to disable it, that the driver does not discover any drives and then we can instruct it to discover certain drives on certain nodes.

Thanks.

**Comment by obacak** [comment]
Thank you for the reply, my colleague will open a ticket about skipping unreadable drives.

I believe I have completely misunderstood what dynamic discovery is and indeed it is not enabled by default, I have figured it out with `install --dry-dun` but then what does it do exactly, I could not find detailed documentation about it in the documents?
